app/lib/zap/controller.js

```diff
@@ -175,52 +180,71 @@ class ZapController {
     return this.startNeutrino()
   }
 
-  onBeforeStartRemoteLnd() {
+  async onBeforeStartRemoteLnd() {
     mainLog.debug('[FSM] onBeforeStartRemoteLnd...')
     mainLog.info('Connecting to lnd:')
     mainLog.info(' > host:', this.lndConfig.host)
     mainLog.info(' > cert:', this.lndConfig.cert && truncate(this.lndConfig.cert))
     mainLog.info(' > macaroon:', this.lndConfig.macaroon && truncate(this.lndConfig.macaroon))
 
-    return this.startLightningWallet().catch(e => {
+    try {
+      await this.startLightningWallet()
+    } catch (e) {
+      /*
+        In the GRPC AIP, there are 2 separate services:
+
+        1) `WalletUnlocker`
+        2) `Lightning`
+
+        The `WalletUnlocker` is used for provisioning and unlocking wallets, and the `Lightning` service
+        is used for interacting with unlocked wallets.
+
+        Before we connect to lnd, we have no way to know what state the wallet is in, and therefore which
+        service we need to connect to. The only way for us to find out is to try making a call against it.
+        Only one of the services is available at any given time
+
+        The `startLightningWallet` call attempts to call the `getInfo` method on the Lightning service in order to
+        verify that it is accessible. If it is not, an error 12 is thrown which is the gRPC code for `UNIMPLEMENTED`
+        which indicates that the requested operation is not implemented or not supported/enabled in the service.
+        See https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/constants.js#L129
+      */
+
+      const isSuccess = code => code === LND_METHOD_UNAVAILABLE
+
+      // first check for the successful case
+      if (isSuccess(e.code)) {
+        // connection successful
+        this.sendMessage('startLndSuccess')
+        return this.startWalletUnlocker()
+      }
+      // error messages to help identify certain type of errors
+      const ERROR_MESSAGES = [
```

Comment:
These error messages are specifically errors related to issues with the macaroon. `MACAROON_ERROR_MESSAGES` might be a better name.